### DISCUSSION TITLE
Is backward link an expensive operation?

### DISCUSSION THREAD

**Discussion by clmnin** [geldata/gel#1898] [Q&A] [answered]
From the doc for [backward link](https://www.edgedb.com/docs/cookbook/links#backward-links), it says

To find all movies that a person is starred in we use a backward link traversal `.<` operator:
```
tutorial>SELECT Person {
    first_name,
    movies := Person.<actors[IS Movie].title,
} FILTER .first_name = 'Ryan';
```
This is how backward link traversal works: EdgeDB fetches every object in the entire database having the field `actors` which is a `Person`. So we narrow down the set of objects to `Movie` and select a title from it.

As you've just read, EdgeDB go through the **entire database**. 

**Question**
So it it optimal to have a property in the `Person` type that links to the `Movie`. Like so

```
type Person {
    required property first_name -> str;
    required property last_name -> str;
    link movie -> Movie;
}
```

**Comment by elprans** [comment, marked as answer]
That piece of documentation is a bit strongly worded.  Backward links do not literally go through the entire database.  In the example above `Person.<actors[IS Movie]` is of the same complexity as `Movie.actors`, because it's the same relation, just traversed in a different direction.

**Reply by clmnin** [reply]
So does that mean when a relationship is made between two types/tables it can - by default - be traversed forward and backward? Without an explicit `link` in the second type/table?

**Reply by elprans** [reply]
Yes.  Moreover, an explicit `link` in the second type is a _different_ link, i.e. there is no "magical" matching of links like you might find in some ORMs.

**Reply by clmnin** [reply]
But doesn't this hide that `Person` type has a link with `Movie` type? In a larger system won't it be harder to figure out all the links a type has?

```edgedb
tutorial> \d Movie

type default::Movie {
    required single link __type__ -> schema::Type {
        readonly := true;
    };
    optional multi link actors -> default::Person;
    required single link director -> default::Person;
    required single property id -> std::uuid {
        readonly := true;
    };
    required single property title -> std::str;
    optional single property year -> std::int64;
};

tutorial> \d Person

type default::Person {
    required single link __type__ -> schema::Type {
        readonly := true;
    };
    required single property first_name -> std::str;
    required single property id -> std::uuid {
        readonly := true;
    };
    required single property last_name -> std::str;
};
```

**Reply by elprans** [reply]
If you use a backward link frequently, you can declare it as a computable, like this:

```
type Person {
    multi link movies := .<actors[IS Movie]
}
```

It's a good point about backward link visibility, we'll consider showing them in verbose `describe` mode.
